features/login/impl/src/main/kotlin/io/element/android/features/login/impl/changeserver/ChangeServerPresenter.kt

```diff
@@ -60,11 +61,15 @@ class ChangeServerPresenter(
                 title = data.title,
                 accountProviderUrl = data.url,
             )
-            authenticationService.setHomeserver(data.url).map {
-                authenticationService.getHomeserverDetails().value!!
-                // Valid, remember user choice
-                accountProviderDataSource.userSelection(data)
-            }.getOrThrow()
+            val details = authenticationService.setHomeserver(data.url).getOrThrow()
+            if (details.supportsOidcLogin.not() && details.supportsPasswordLogin.not()) {
+                // Unsupported homeserver
+                throw ChangeServerError.Error(
```

Comment:
I'd remove this `Error` which is a bit confusing, and create individual error for all the defined errors + one Unknown case for the default behavior